internal/trie/node/copy.go

```diff
@@ -4,17 +4,29 @@
 package node
 
 // Copy deep copies the branch.
-func (b *Branch) Copy() Node {
+// Setting copyChildren to true will deep copy
+// children as well.
+func (b *Branch) Copy(copyChildren bool) Node {
 	b.RLock()
 	defer b.RUnlock()
 
 	cpy := &Branch{
-		Children:   b.Children, // copy interface pointers
-		dirty:      b.dirty,
+		Dirty:      b.Dirty,
 		generation: b.generation,
 	}
 	copy(cpy.Key, b.Key)
 
+	if copyChildren {
+		for i, child := range b.Children {
+			if child == nil {
+				continue
+			}
+			cpy.Children[i] = child.Copy(copyChildren)
+		}
+	} else {
+		cpy.Children = b.Children // copy interface pointers only
+	}
+
 	if b.Key != nil {
 		cpy.Key = make([]byte, len(b.Key))
 		copy(cpy.Key, b.Key)
```

Comment:
this condition executes the same code already executed at line `16` unless by the allocation before copy, maybe remove the code at line `16`

Reply:
Not sure what you mean, there is no code on line 16 really :thinking: 
Also note this is an array so no need to allocate it if that's what you mean?

Reply:
Some lines above (before the `if copyChildren`) there is a `copy(cpy.Key, b.Key)` and here we check `if b.Key != nil` then do the copy again

Reply:
Totally what was that, anyways; fixed!